pkg/controller/emrcontainers/virtualcluster/setup.go

```diff
@@ -0,0 +1,162 @@
+package virtualcluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	svcsdk "github.com/aws/aws-sdk-go/service/emrcontainers"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/connection"
+	"github.com/crossplane/crossplane-runtime/pkg/controller"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	cpresource "github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/pkg/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	svcapitypes "github.com/crossplane-contrib/provider-aws/apis/emrcontainers/v1alpha1"
+	"github.com/crossplane-contrib/provider-aws/apis/v1alpha1"
+	awsclients "github.com/crossplane-contrib/provider-aws/pkg/clients"
+	"github.com/crossplane-contrib/provider-aws/pkg/features"
+)
+
+const (
+	firstObserveID    = "invalid"
+	terminatedMessage = "VirtualCluster is already terminated"
+	errListTag        = "cannot list tags"
+	errUntag          = "cannot remove tags"
+	errTag            = "cannot add tags"
+)
+
+// SetupVirtualCluster adds a controller that reconciles VirtualCluster.
+func SetupVirtualCluster(mgr ctrl.Manager, o controller.Options) error {
+	name := managed.ControllerName(svcapitypes.VirtualClusterGroupKind)
+	opts := []option{
+		func(e *external) {
+			e.isUpToDate = isUpToDate
+			e.preCreate = preCreate
+			e.preObserve = preObserve
+			e.postCreate = postCreate
+			e.postDelete = postDelete
+			e.postObserve = postObserve
+			e.update = e.updater
+		},
+	}
+
+	cps := []managed.ConnectionPublisher{managed.NewAPISecretPublisher(mgr.GetClient(), mgr.GetScheme())}
+	if o.Features.Enabled(features.EnableAlphaExternalSecretStores) {
+		cps = append(cps, connection.NewDetailsManager(mgr.GetClient(), v1alpha1.StoreConfigGroupVersionKind))
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(o.ForControllerRuntime()).
+		For(&svcapitypes.VirtualCluster{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(svcapitypes.VirtualClusterGroupVersionKind),
+			managed.WithExternalConnecter(&connector{kube: mgr.GetClient(), opts: opts}),
+			managed.WithPollInterval(o.PollInterval),
+			managed.WithLogger(o.Logger.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))),
+			managed.WithConnectionPublishers(cps...)))
+}
+
+func preObserve(ctx context.Context, cr *svcapitypes.VirtualCluster, input *svcsdk.DescribeVirtualClusterInput) error {
+	externalName := meta.GetExternalName(cr)
+	if externalName == cr.Name {
```

Comment:
@nabuskey why you need this it block?

Reply:
EMR containers does input parameter validation and return `ValidationException` if the ID does not satisfy `[0-9a-z]+`. This is a problem when observing a VirtualCluster or a JobRun for the first time because external name is set to the CR's name and CR names could potentially do not satisfy the regex expression. So this is done to ensure API returns 404 during observe for the first time.
It looks like you can't return non-404 error during observe, otherwise this is hit: https://github.com/crossplane/crossplane-runtime/blob/65044f043902ebb1ab1e1bb8685c3208f1c301d2/pkg/reconciler/managed/reconciler.go#L781
I thought about setting the generator to look for ValidationException as 404, but it didn't feel right.